LibreHardwareMonitorLib/Hardware/Battery/BatteryGroup.cs

```diff
@@ -0,0 +1,231 @@
+        private static int GetCharSize(CharSet charSet = CharSet.Auto) => charSet == CharSet.Auto ? Marshal.SystemDefaultCharSize : (charSet == CharSet.Unicode ? 2 : 1);
+        private static bool IsValue(IntPtr ptr) => ptr.ToInt64() >> 16 == 0;
+
+        private static string GetString(IntPtr ptr, CharSet charSet = CharSet.Auto, long allocatedBytes = long.MaxValue)
+        {
+            if (IsValue(ptr)) return null;
+            var sb = new StringBuilder();
+            unsafe
+            {
+                var chkLen = 0L;
+                if (GetCharSize(charSet) == 1)
+                {
+                    for (var uptr = (byte*)ptr; chkLen < allocatedBytes && *uptr != 0; chkLen++, uptr++)
+                        sb.Append((char)*uptr);
+                }
+                else
+                {
+                    for (var uptr = (ushort*)ptr; chkLen + 2 <= allocatedBytes && *uptr != 0; chkLen += 2, uptr++)
+                        sb.Append((char)*uptr);
+                }
+            }
+            return sb.ToString();
+        }
+
+        public BatteryGroup(ISettings settings)
+        {
+            // No implementation for battery information on Unix systems
+            if (Software.OperatingSystem.IsUnix)
+            {
+                return;
+            }
+
+            IntPtr hdev = SetupApi.SetupDiGetClassDevs(ref SetupApi.GUID_DEVICE_BATTERY, IntPtr.Zero, IntPtr.Zero, SetupApi.DIGCF_PRESENT | SetupApi.DIGCF_DEVICEINTERFACE);
+            if (SetupApi.INVALID_HANDLE_VALUE != hdev)
+            {
+                // Limit battery search to 100 batteries
+                for (uint idev = 0; idev < 100; idev++)
+                {
+                    SetupApi.SP_DEVICE_INTERFACE_DATA did = default;
+                    did.cbSize = (uint)Marshal.SizeOf(typeof(SetupApi.SP_DEVICE_INTERFACE_DATA));
+
+                    if (SetupApi.SetupDiEnumDeviceInterfaces(hdev,
+                                                             IntPtr.Zero,
+                                                             ref SetupApi.GUID_DEVICE_BATTERY,
+                                                             idev,
+                                                             ref did))
+                    {
+                        
+                        SetupApi.SetupDiGetDeviceInterfaceDetail(hdev,
+                                                                 did,
+                                                                 IntPtr.Zero,
+                                                                 0,
+                                                                 out uint cbRequired,
+                                                                 IntPtr.Zero);
+                        if (Marshal.GetLastWin32Error() == SetupApi.ERROR_INSUFFICIENT_BUFFER)
+                        {
+                            IntPtr pdidd = Kernel32.LocalAlloc(Kernel32.LPTR, cbRequired);
```

Comment:
Marshal.AllocHGlobal and below Marshal.FreeHGlobal?

Reply:
No. Because it IS necessary (in this case) to pass the LPTR (0x0000 | 0x0040) flag in LocalAlloc(). Marshal.AllocHGlobal() calls LocalAlloc(), but it passes a LMEM_FIXED (0x0000) flag. If I use the Marshal.AllocHGlobal() method, LHM crashes inmediately.